README.md
Murpoint
========

Introduction
------------
This tool was created to harvest all RDF from a specific domain. You can provide
it with a starting URI, and it will follow all links that match the hostname of
the starting URI, harvesting everything to a text file containing n-triples. No
URIs outside of the starting host will be accessed.

Installation
------------
This project uses Composer. Run "composer install" to set up dependencies.

Basic Usage
-----------
php Murpoint.php [starting URI] [output file]

Resume Functionality
--------------------
If Murpoint encounters a fatal error during execution, it will dump its state to a file. You can attempt to retry the crawl later with:

php Murpoint.php --resume [name of dump file]

Additional Options
------------------
You can obtain information on all available options with:

php Murpoint.php --help

Background
----------
This project was designed to feed the triple store used by dimenovels.org, and
is named for Captain Howard Murpoint, the villain of Charles Garvice's dime
novel, "The Spider and the Fly."


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 3.1 - 2020-05-18

### Added

- Option to set HTTP timeout (--timeout=n)

### Changed

- Error log messages now include more details.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 3.0 - 2020-05-01

### Added

- --log option for logging to file instead of console.
- Progress bar.

### Changed

- Refactored to use Symfony\Console framework.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Bug: last URL accessed was dropped when resuming interrupted job.

## 2.0.1 - 2020-03-11

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Bug: output filename was not defined when handling errors during a resumed job.

## 2.0 - 2019-06-19

### Added

- State save on exception, with --resume option.

### Changed

- Constructor parameters for Crawler class have changed to support new functionality.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 1.0 - 2016-08-01

Initial release.

Murpoint.php
<?php
/**
 * Murpoint - A simple RDF harvesting tool.
 *
 * PHP version 5
 *
 * Copyright (C) Villanova University 2016.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author   Demian Katz <demian.katz@villanova.edu>
 * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
 * @link     https://github.com/FalveyLibraryTechnology/Murpoint
 */
require 'vendor/autoload.php';

$command = new \Murpoint\Command();
$application = new \Symfony\Component\Console\Application();
$application->add($command);
$application->setDefaultCommand($command->getName(), true);
return $application->run();


src/Crawler.php
<?php
/**
 * Murpoint - A simple RDF harvesting tool.
 *
 * PHP version 5
 *
 * Copyright (C) Villanova University 2016.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author   Demian Katz <demian.katz@villanova.edu>
 * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
 * @link     https://github.com/FalveyLibraryTechnology/Murpoint
 */
namespace Murpoint;

use EasyRdf\Graph;

/**
 * Main crawler class to harvest RDF.
 */
class Crawler
{
    /**
     * File handle for output.
     *
     * @var resource
     */
    protected $handle;

    /**
     * Host to harvest.
     *
     * @var string
     */
    protected $host;

    /**
     * Name of output file.
     *
     * @var string
     */
    protected $outFile;

    /**
     * Queue of pending URLs to fetch.
     *
     * @var string[]
     */
    protected $queue;

    /**
     * List of previously-visited URLs.
     *
     * @var string[]
     */
    protected $visited;

    /**
     * Callback for logging.
     *
     * @var \Callable
     */
    protected $logCallback = null;

    /**
     * Callbacks for progress bar.
     *
     * @var array
     */
    protected $barCallbacks = [];

    /**
     * Constructor
     *
     * Note: The last three parameters are only intended to be used when resuming
     * from a saved state created during a failed harvest.
     *
     * @param string $host    Hostname within which to restrict fetching
     * @param string $outFile Output file for triples
     * @param array  $queue   Queue (default = empty)
     * @param array  $visited Visited URLs (default = empty)
     * @param string $mode    File write mode (default = 'w')
     */
    public function __construct($host, $outFile, $queue = [], $visited = [], $mode = 'w')
    {
        $this->host = $host;
        $this->handle = fopen($outFile, $mode);
        $this->outFile = $outFile;
        if (!$this->handle) {
            throw new \Exception('Cannot open ' . $outFile . ' for writing.');
        }
        $this->queue = $queue;
        $this->visited = $visited;
    }

    /**
     * Set a log callback.
     *
     * @param \Callable $callback Logging callback
     *
     * @return void
     */
    public function setLogCallback($callback)
    {
        $this->logCallback = $callback;
    }

    /**
     * Set callbacks for the progress bar.
     *
     * @param \Callable $progressCallback Advance progress by one step
     * @param \Callable $maxStepsCallback Advance maximum size by one step
     *
     * @return void
     */
    public function setProgressBarCallbacks($progressCallback, $maxStepsCallback)
    {
        $this->barCallbacks['progress'] = $progressCallback;
        $this->barCallbacks['maxSteps'] = $maxStepsCallback;
    }

    /**
     * Trigger a progress bar action
     *
     * @param string $type Type of action to trigger
     *
     * @return void
     */
    protected function updateProgressBar($type)
    {
        if (isset($this->barCallbacks[$type])) {
            call_user_func($this->barCallbacks[$type]);
        }
    }

    /**
     * Log a message.
     *
     * @param string $msg Message to log
     *
     * @return void
     */
    protected function log($msg)
    {
        if ($this->logCallback) {
            call_user_func($this->logCallback, $msg);
        } else {
            echo $msg . "\n";
        }
    }

    /**
     * Add a URL to the queue.
     *
     * @param string $url URL
     *
     * @return void
     */
    public function enqueue($uri)
    {
        // strip off fragment identifiers:
        $uri = current(explode('#', $uri));

        // make sure the hostname matches:
        $urlParts = parse_url($uri);

        if (!in_array($uri, $this->visited) && !in_array($uri, $this->queue) && $urlParts['host'] == $this->host) {
            $this->log("Queueing $uri");
            $this->updateProgressBar('maxSteps');
            $this->queue[] = $uri;
        }
    }

    /**
     * Run the harvest.
     *
     * @return void
     */
    public function run()
    {
        while (!empty($this->queue)) {
            $this->processUri(array_shift($this->queue));
        }
    }

    /**
	 * Get the output filename.
	 *
	 * @return string
	 */
	public function getOutputFilename()
	{
	    return $this->outFile;
	}

    /**
     * Save the current state of the crawler to disk.
     *
     * @param string $filename Output file name.
     *
     * @return void
     */
    public function saveState($filename)
    {
        $state = [
            'host' => $this->host,
            'outFile' => $this->outFile,
            'queue' => $this->queue,
            'visited' => $this->visited,
        ];
        file_put_contents($filename, serialize($state));
    }

    /**
     * Create a new Crawler from a saved state file.
     *
     * @param string $filename State to load
     *
     * @return Crawler
     */
    public static function fromState($filename)
    {
        $state = unserialize(file_get_contents($filename));
        return new Crawler(
            $state['host'], $state['outFile'],
            $state['queue'], $state['visited'], 'a'
        );
    }

    /**
     * Internal support method: process the contents of a URI.
     *
     * @param string $uri URI to process.
     *
     * @return void
     */
    protected function processUri($uri)
    {
        $this->log("Accessing $uri...");
        $retries = 3;
        while (true) {
            try {
                $graph = Graph::newAndLoad($uri);
                break;
            } catch (\Exception $e) {
                $retries--;
                if ($retries < 0) {
                    // Throw the bad URL back on the queue for later retry...
                    $this->enqueue($uri);
                    throw $e;
                }
                $this->log(
                    "Error: " . $e->getMessage() . "\n"
                    . "Sleeping for 10 secs, then retrying ($retries tries left)..."
                );
                sleep(10);
            }
        }
        $triples = $graph->serialise('ntriples');
        fwrite($this->handle, $triples);
        $this->visited[] = $uri;
        $this->updateProgressBar('progress');
        preg_match_all('/<(https?:\/\/[^>]*)>/', $triples, $matches);
        foreach (array_unique($matches[1]) as $uri) {
            $this->enqueue($uri);
        }
    }
}


src/Command.php
<?php
/**
 * Murpoint - A simple RDF harvesting tool.
 *
 * PHP version 5
 *
 * Copyright (C) Villanova University 2016.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * @author   Demian Katz <demian.katz@villanova.edu>
 * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
 * @link     https://github.com/FalveyLibraryTechnology/Murpoint
 */
namespace Murpoint;

use Symfony\Component\Console\Helper\ProgressBar;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;

/**
 * Symfony command to run Murpoint.
 */
class Command extends \Symfony\Component\Console\Command\Command
{
    /**
     * The name of the command
     *
     * @var string
     */
	protected static $defaultName = 'murpoint';

    /**
     * Configure the command.
     *
     * @return void
     *
     * @SuppressWarnings(PHPMD.ExcessiveMethodLength)
     */
    protected function configure()
    {
        $this
            ->setDescription('Murpoint crawler')
            ->setHelp('Harvests RDF over HTTP.')
            ->addArgument(
                'url',
                InputArgument::OPTIONAL,
				'The starting URL to harvest from'
            )->addArgument(
                'output',
                InputArgument::OPTIONAL,
                'The file used for storing harvested RDF'
            )->addOption(
                'resume',
                null,
                InputOption::VALUE_REQUIRED,
                'Resume an interrupted harvest using a dump file'
			)->addOption(
				'log',
				null,
				InputOption::VALUE_REQUIRED,
				'Filename for logging (omit to use console)'
			)->addOption(
				'timeout',
				null,
				InputOption::VALUE_REQUIRED,
				'HTTP timeout (in seconds)'
			);
		}

	/**
	 * Get the crawler object based on input parameters.
	 *
     * @param InputInterface  $input  Input object
     * @param OutputInterface $output Output object
	 *
	 * @return Crawler
	 * @throws \Exception
	 */
	protected function getCrawlerFromInput(InputInterface $input, OutputInterface $output): Crawler
	{
		// Special case: resume existing crawl
		$resume = $input->getOption('resume');
		if ($resume !== null) {
			if (empty($resume) || !file_exists($resume)) {
				throw new \Exception("Cannot open resume file.");
			}
			$crawler = Crawler::fromState($resume);
			$crawler->setLogCallback(
				$this->getLogCallback($input->getOption('log'), $output)
			);
			return $crawler;
		}
		// Standard case: starting new crawl
		$url = $input->getArgument('url');
		$urlParts = empty($url) ? [] : parse_url($url);
		$outFile = $input->getArgument('output');
		if (!isset($urlParts['host']) || empty($urlParts['host'])) {
			throw new \Exception(
				"First parameter must be valid starting URL."
			);
		}
		if (!$outFile) {
			throw new \Exception(
				"Second parameter must be writable file for output."
			);
		}
		$crawler = new Crawler($urlParts['host'], $outFile);
		$crawler->setLogCallback(
			$this->getLogCallback($input->getOption('log'), $output)
		);
		$crawler->enqueue($url);
		return $crawler;
	}

	/**
	 * Get the callback for logging messages from Murpoint.
	 *
	 * @param string          $filename Log filename (or false/null to use console)
     * @param OutputInterface $output   Output object
	 *
	 * @return \Callable
	 */
	protected function getLogCallback($filename, $output)
	{
		// No filename? Use console.
		if (!$filename) {
			return function ($msg) use ($output) {
				$output->writeln($msg);
			};
		}
		// Filename set? Use file logging.
		$handle = fopen($filename, 'a');
		register_shutdown_function(
			function () use ($handle) {
				fclose($handle);
			}
		);
		return function ($msg) use ($handle) {
			fputs($handle, $msg . "\n");
		};
	}

    /**
     * Run the command.
     *
     * @param InputInterface  $input  Input object
     * @param OutputInterface $output Output object
     *
     * @return int 0 for success
     */
    protected function execute(InputInterface $input, OutputInterface $output)
	{
		// Set up HTTP timeout, as needed:
		if ($timeout = $input->getOption('timeout')) {
			\EasyRdf\Http::getDefaultHttpClient()->setConfig(compact('timeout'));
		}
		// Special case: resuming from dump file:
		try {
			$crawler = $this->getCrawlerFromInput($input, $output);
			$progressBar = new ProgressBar($output, 0);
			$progressCallback = function () use ($progressBar) {
				$progressBar->advance();
			};
			$maxStepsCallback = function () use ($progressBar) {
				static $steps = 0;
				$steps++;
				$progressBar->setMaxSteps($steps);
			};
			$crawler->setProgressBarCallbacks($progressCallback, $maxStepsCallback);
			$progressBar->start();
		} catch (\Exception $e) {
			$output->writeln($e->getMessage());
			return 1;
		}

		try {
			$crawler->run();
		} catch (\Exception $e) {
			$output->writeln("Fatal error: " . $e->getMessage());
			$output->writeln("Dumping state...");
			$dumpFile = $crawler->getOutputFilename() . '.state';
			$crawler->saveState($dumpFile);
			$output->writeln(
				"You can attempt to resume later with: php Murpoint.php --resume "
				. $dumpFile
			);
			return 1;
		}
		$progressBar->finish();
		return 0;
	}
}
